proposals/js-string-builtins/Overview.md

```diff
@@ -28,221 +28,277 @@ The first problem is that existing API’s require a calling convention conversi
-  }
-  return result;
-}
-```
-
-[4]: "array[i]" is meant to express "load the i-th element of the array", in Wasm terms: (array.get_u $i16-array-type (local.get $array) (local.get $i)) for an appropriate $i16-array-type.
-
-Note: This function takes an i16 array defined in its own recursion group. If this is an issue for a toolchain, we can look into how to relax the function type while still maintaining good performance.
+  // NOTE: `start` and `end` are interpreted as signed 32-bit integers when
+  // converted to JS values using standard conversions. Reinterpret them as
+  // unsigned here.
+  start >>>= 0;
+  end >>>= 0;
 
-### "wasm:js-string" "fromWtf8Array"
-
-```
-func fromWtf8Array(
-  array: (ref null (array i8)),
-  start: i32,
-  end: i32
-) -> (ref extern)
-{
-  start = ToUint32(start);
-  end = ToUint32(end);
+  if (array === null)
+    trap();
 
-  // [2]
-  if (end > array.length || start > end)
-    trap;
+  if (start > end ||
+      end > array_len(array))
+    trap();
 
-  // This summarizes as: "decode the WTF-8 string stored at array[start:end],
-  // or trap if there is no valid WTF-8 string there".
   let result = "";
-  while (start < end) {
-    if there is no valid wtf8-encoded code point at array[start]
-      trap;
-    let c be the code point at array[start];
-    // [1]
-    result += String.fromCodePoint(c);
-    increment start by as many bytes as it took to decode c;
+  for(let i = start; i < end; i++) {
+    let charCode = array_i16_get(array, i);
+    result += String.fromCharCode(charCode);
   }
   return result;
 }
 ```
 
-Note to implementers: while this is the only usage of WTF-8 in this document, it shouldn't be very burdensome to implement, because all existing strings in Wasm modules (import/export names, contents of the name section) are already in UTF-8 format, so implementations must already have decoding infrastructure for that. We need the relaxation from UTF-8 to WTF-8 to support WTF-16 based source languages, which may have unpaired surrogates in string constants in existing/legacy code.
-
-### "wasm:js-string" "toWtf16Array"
-
-"start" is the index in the array where the first codeunit of the string will be written.
-
-Returns the number of codeunits written. Traps if the string doesn't fit into the array.
+### "wasm:js-string" "copyToCharCodeArray"
 
 ```
-func toWtf16Array(
+/// Copy a string into a pre-allocated mutable i16 array at `start` index.
+///
+/// Returns the number of char codes written, which is equal to the length of
+/// the string.
+///
+/// Traps if the string doesn't fit into the array.
+func copyToCharCodeArray(
   string: externref,
   array: (ref null (array (mut i16))),
   start: i32
 ) -> i32
 {
-  if (typeof string !== "string")
-    trap;
+  // NOTE: `start` is interpreted as a signed 32-bit integer when converted
+  // to a JS value using standard conversions. Reinterpret as unsigned here.
+  start >>>= 0;
+
+  if (array === null)
```

Comment:
I was thinking that doing the upfront bounds check would allow for the most flexibility in implementation by making sure any re-ordering of the loop wouldn't need to preserve the proper access order. But I'm open to filing an issue to discuss this more.